evm/src/cpu/kernel/asm/ecrecover.asm

```diff
@@ -0,0 +1,193 @@
+    JUMPDEST
+    // stack: hash, v, r, s, retdest
+
+    // Check if inputs are valid.
+    %ecrecover_input_check
+    // stack: isValid(v,r,s), hash, v, r, s, retdest
+
+    // Lift r to an elliptic curve point if possible.
+    SWAP2
+    // stack: v, hash, isValid(v,r,s), r, s, retdest
+    DUP4
+    // stack: r, v, hash, isValid(v,r,s), r, s, retdest
+    %secp_lift_x
+    // stack: y, sqrtOk, hash, isValid(v,r,s), r, s, retdest
+
+    // If inputs are invalid or lifting fails, abort.
+    SWAP3
+    // stack: isValid(v,r,s), sqrtOk, hash, y, r, s, retdest
+    AND
+    // stack: isValid(v,r,s) & sqrtOk, hash, y, r, s, retdest
+    %jumpi(ecrecover_valid_input)
+    // stack: hash, y, r, s, retdest
+    %pop4
+    // stack: retdest
+    %ecrecover_invalid_input
+
+// ecrecover precompile.
+// Assumption: Inputs are valid.
+// Pseudo-code:
+// let P = lift_x(r, recovery_id);
+// let r_inv = r.inverse();
+// let u1 = s * r_inv;
+// let u2 = -hash * r_inv;
+// return u1*P + u2*GENERATOR;
+ecrecover_valid_input:
+    JUMPDEST
+    // stack: hash, y, r, s, retdest
+
+    // Compute u1 = s * r^(-1)
+    SWAP1
+    // stack: y, hash, r, s, retdest
+    DUP3
+    // stack: r, y, hash, x, s, retdest (r=x)
+    %inverse_secp_scalar
+    // stack: r^(-1), y, hash, x, s, retdest
+    DUP1
+    // stack: r^(-1), r^(-1), y, hash, x, s, retdest
+    SWAP5
+    // stack: s, r^(-1), y, hash, x, r^(-1), retdest
+    %mulmodn_secp_scalar
+    // stack: u1, y, hash, x, r^(-1), retdest
+
+
+    // Compute (X,Y) = u1 * (x,y)
+    PUSH ecrecover_with_first_point
+    // stack: ecrecover_with_first_point, u1, y, hash, x, r^(-1), retdest
+    SWAP1
+    // stack: u1, ecrecover_with_first_point, y, hash, x, r^(-1), retdest
+    SWAP2
+    // stack: y, ecrecover_with_first_point, u1, hash, x, r^(-1), retdest
+    SWAP1
+    // stack: ecrecover_with_first_point, y, u1, hash, x, r^(-1), retdest
+    SWAP3
+    // stack: hash, y, u1, ecrecover_with_first_point, x, r^(-1), retdest
+    SWAP4
+    // stack: x, y, u1, ecrecover_with_first_point, hash, r^(-1), retdest
+    %jump(ec_mul_valid_point_secp)
+
+// ecrecover precompile.
+// Assumption: (X,Y) = u1 * P. Result is (X,Y) + u2*GENERATOR
+ecrecover_with_first_point:
+    JUMPDEST
+    // stack: X, Y, hash, r^(-1), retdest
+    %secp_scalar
+    // stack: p, X, Y, hash, r^(-1), retdest
+    SWAP1
+    // stack: X, p, Y, hash, r^(-1), retdest
+    SWAP4
+    // stack: r^(-1), p, Y, hash, X, retdest
+    SWAP2
+    // stack: Y, p, r^(-1), hash, X, retdest
+    SWAP3
+    // stack: hash, p, r^(-1), Y, X, retdest
+
+    // Compute u2 = -hash * r^(-1)
+    MOD
+    // stack: hash%p, r^(-1), Y, X, retdest
+    %secp_scalar
+    // stack: p, hash%p, r^(-1), Y, X, retdest
+    SUB
+    // stack: -hash, r^(-1), Y, X, retdest
+    %mulmodn_secp_scalar
+    // stack: u2, Y, X, retdest
+
+    // Compute u2 * GENERATOR and chain the call to `ec_mul` with a call to `ec_add` to compute PUBKEY = (X,Y) + u2 * GENERATOR,
```

Comment:
So here `PUBKEY` is computed like 
```
(X,Y) = u1 * P
(X', Y') = u2 * GENERATOR
PUBKEY = (X,Y) + (X',Y')
```
It would probably be more efficient to compute this using a MSM function. If we end up merging the `ec_mul` functions for Secp256k1 and BN254 then the MSM function would add a bunch of kernel code, so the tradeoffs are a bit unclear for now.

Reply:
Yeah seems fine for now 👍